File: services/gateway/gateway-api/src/main/java/com/example/gateway/api/gateway/TradeLogGateway.java
package com.example.gateway.api.gateway;

import com.example.gateway.api.spec.model.ShareJournalGWModel;
import com.example.tradelog.api.spec.model.DividendJournalModel;
import com.example.tradelog.api.spec.model.DividendTransactionsResponse;
import com.example.tradelog.api.spec.model.OptionJournalModel;
import com.example.tradelog.api.spec.model.OptionTransactionsResponse;
import com.example.tradelog.api.spec.model.ShareJournalModel;
import com.example.tradelog.api.spec.model.ShareTransactionsResponse;
import com.example.tradelog.api.spec.model.TradeSummaryModel;
import com.example.tradelog.api.spec.model.TransactionSettingsModel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.List;
import java.util.concurrent.CompletableFuture;

@Service
public class TradeLogGateway {

    private static final Logger log = LoggerFactory.getLogger(TradeLogGateway.class);

    private final RestTemplate restTemplate;
    private final String API_URL;

    public TradeLogGateway(RestTemplate restTemplate, @Value("${gateway.tradelog.url}") String url) {
        this.restTemplate = restTemplate;
        this.API_URL = url;
    }


    public List<String> getAllTradedSymbols(String accountId) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/transactions/symbols");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        ResponseEntity<List<String>> responseEntity = restTemplate
                .exchange(builder.build(""), HttpMethod.GET, new HttpEntity<>(headers), new ParameterizedTypeReference<List<String>>() {});
        return responseEntity.getBody();
    }


    @Async("asyncExecutor")
    public CompletableFuture<ShareTransactionsResponse> getShareTransactionsBySymbol(String accountId, String symbol) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/shares/{symbol}");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        ResponseEntity<ShareTransactionsResponse> responseEntity =
                restTemplate.exchange(builder.build(symbol), HttpMethod.GET, new HttpEntity<>(headers), ShareTransactionsResponse.class);
        return CompletableFuture.completedFuture(responseEntity.getBody());
    }


    @Async("asyncExecutor")
    public CompletableFuture<OptionTransactionsResponse> getOptionTransactionsBySymbol(String accountId, String symbol) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/options/{symbol}");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        ResponseEntity<OptionTransactionsResponse> responseEntity =
                restTemplate.exchange(builder.build(symbol), HttpMethod.GET, new HttpEntity<>(headers), OptionTransactionsResponse.class);
        return CompletableFuture.completedFuture(responseEntity.getBody());
    }


    @Async("asyncExecutor")
    public CompletableFuture<DividendTransactionsResponse> getDividendTransactionsBySymbol(String accountId, String symbol) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/dividends/{symbol}");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        ResponseEntity<DividendTransactionsResponse> responseEntity =
                restTemplate.exchange(builder.build(symbol), HttpMethod.GET, new HttpEntity<>(headers), DividendTransactionsResponse.class);
        return CompletableFuture.completedFuture(responseEntity.getBody());
    }


    public ShareJournalModel createShareTrade(String accountId, ShareJournalModel model) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/shares");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        HttpEntity<ShareJournalModel> request = new HttpEntity<>(model, headers);
        ResponseEntity<ShareJournalModel> responseEntity = restTemplate.exchange(builder.build(model.getTransactionDetails().getSymbol()), HttpMethod.POST, request, ShareJournalModel.class);
        return responseEntity.getBody();
    }

    public OptionJournalModel createOptionTrade(String accountId, OptionJournalModel model) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/options");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        HttpEntity<OptionJournalModel> request = new HttpEntity<>(model, headers);
        ResponseEntity<OptionJournalModel> responseEntity = restTemplate.exchange(builder.build(model.getTransactionDetails().getSymbol()), HttpMethod.POST, request, OptionJournalModel.class);
        return responseEntity.getBody();
    }

    public DividendJournalModel createDividendTrade(String accountId, DividendJournalModel model) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/dividends");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        HttpEntity<DividendJournalModel> request = new HttpEntity<>(model, headers);
        ResponseEntity<DividendJournalModel> responseEntity = restTemplate.exchange(builder.build(model.getTransactionDetails().getSymbol()), HttpMethod.POST, request, DividendJournalModel.class);
        return responseEntity.getBody();
    }

    public void updateShareTrade(String accountId, String transactionId, ShareJournalModel model) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/shares/{id}");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        HttpEntity<ShareJournalModel> request = new HttpEntity<>(model, headers);
        restTemplate.exchange(builder.build(transactionId), HttpMethod.PUT, request, Object.class);
    }

    public void deleteShareTrade(String accountId, String transactionId) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/shares/{id}");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        restTemplate.exchange(builder.build(transactionId), HttpMethod.DELETE, new HttpEntity(headers), Object.class);
    }

    public void deleteOptionTrade(String accountId, String transactionId) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/options/{id}");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        restTemplate.exchange(builder.build(transactionId), HttpMethod.DELETE, new HttpEntity(headers), Object.class);
    }

    public void deleteDividendTrade(String accountId, String transactionId) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/dividends/{id}");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        restTemplate.exchange(builder.build(transactionId), HttpMethod.DELETE, new HttpEntity(headers), Object.class);
    }

    public List<TradeSummaryModel> getSummary(String accountId) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/transactions/summary");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        ResponseEntity<List<TradeSummaryModel>> responseEntity = restTemplate
                .exchange(builder.build(""), HttpMethod.GET, new HttpEntity<>(headers), new ParameterizedTypeReference<List<TradeSummaryModel>>() {});
        return responseEntity.getBody();
    }

    public void updateTransactionSettings(String accountId, String transactionId, TransactionSettingsModel optionsModel) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/transactions/options/{id}");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        HttpEntity<TransactionSettingsModel> request = new HttpEntity<>(optionsModel, headers);
        restTemplate.exchange(builder.build(transactionId), HttpMethod.PUT, request, Object.class);
    }

    public void updateTransactionSettingsBulk(String accountId, List<TransactionSettingsModel> models) {
        UriComponentsBuilder builder = UriComponentsBuilder
                .fromHttpUrl(API_URL)
                .path("/transactions/settings/bulk");

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("accountId", accountId);

        HttpEntity<List<TransactionSettingsModel>> request = new HttpEntity<>(models, headers);
        restTemplate.exchange(builder.build(""), HttpMethod.PUT, request, Object.class);
    }
}

File: services/gateway/gateway-api/src/main/kotlin/com/example/gateway/api/rest/gateway/TradeLogGateway.kt
package com.example.gateway.api.rest.gateway

import arrow.core.Either
import com.example.common.exception.ApiException
import com.example.common.exception.ApiExceptionCode
import com.example.gateway.api.core.model.*
import com.example.gateway.api.rest.converter.*
import com.example.tradelog.api.spec.model.*
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClientException
import org.springframework.web.client.RestTemplate
import org.springframework.web.util.UriComponentsBuilder
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.stream.Collectors

@Service
class TradeLogGateway(
        private val restTemplate: RestTemplate,
        @Value("\${gateway.tradelog.url}") private val url: String
) {

    companion object {
        private const val PROTOBUF_MEDIA_TYPE_VALUE = "application/x-protobuf"
        private const val PORTFOLIO_ID_PARAM_NAME = "portfolio-id"
        private const val ACCOUNT_ID_HEADER_NAME = "x-account-id"
    }

    fun getAllActiveSymbols(): Either<ApiException, List<String>> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/transactions/active-symbols")
        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)

        return try {
            val responseEntity = restTemplate
                    .exchange(builder.build("").toString(), HttpMethod.GET, HttpEntity<Any>(headers), TLActiveSymbolsResponse::class.java)

            val dto: TLActiveSymbolsResponse? = responseEntity.body

            return if (dto != null) {
                Either.Right(ActiveSymbolsResponseConverter.toModel(dto))
            } else {
                Either.Right(Collections.emptyList())
            }
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }
    }

    @Async("asyncExecutor")
    fun getAllShareTransactions(supportMetadata: SupportMetadata, symbol: String): CompletableFuture<List<ShareJournalModel>> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/shares/{symbol}")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        val responseEntity = restTemplate
                .exchange(builder.build(symbol).toString(), HttpMethod.GET, HttpEntity<Any>(headers), TLShareTransactionsResponse::class.java)

        val dto: TLShareTransactionsResponse? = responseEntity.body

        return if (dto != null) {
            CompletableFuture.completedFuture(
                    dto.shareItemsList.stream().map { ShareJournalConverter.toModel(it) }.collect(Collectors.toList())
            )
        } else {
            CompletableFuture.completedFuture(Collections.emptyList())
        }
    }

    @Async("asyncExecutor")
    fun getAllOptionTransactions(supportMetadata: SupportMetadata, symbol: String): CompletableFuture<List<OptionJournalModel>> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/options/{symbol}/")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        val responseEntity = restTemplate
                .exchange(builder.build(symbol).toString(), HttpMethod.GET, HttpEntity<Any>(headers), TLOptionTransactionsResponse::class.java)

        val dto: TLOptionTransactionsResponse? = responseEntity.body

        return if (dto != null) {
            CompletableFuture.completedFuture(
                    dto.optionItemsList.stream().map { OptionJournalConverter.toModel(it) }.collect(Collectors.toList())
            )
        } else {
            CompletableFuture.completedFuture(Collections.emptyList())
        }
    }

    @Async("asyncExecutor")
    fun getAllDividendTransactions(supportMetadata: SupportMetadata, symbol: String): CompletableFuture<List<DividendJournalModel>> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/dividends/{symbol}/")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        val responseEntity = restTemplate
                .exchange(builder.build(symbol).toString(), HttpMethod.GET, HttpEntity<Any>(headers), TLDividendTransactionsResponse::class.java)

        val dto: TLDividendTransactionsResponse? = responseEntity.body

        return if (dto != null) {
            CompletableFuture.completedFuture(
                    dto.dividendItemsList.stream().map { DividendJournalConverter.toModel(it) }.collect(Collectors.toList())
            )
        } else {
            CompletableFuture.completedFuture(Collections.emptyList())
        }
    }

    fun getSummary(supportMetadata: SupportMetadata): Either<ApiException, List<TradeSummaryModel>> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/transactions/summary")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        return try {
            val responseEntity = restTemplate
                    .exchange(builder.build("").toString(), HttpMethod.GET, HttpEntity<Any>(headers), TLTradeSummaryResponse::class.java)

            val dto: TLTradeSummaryResponse? = responseEntity.body

            return if (dto != null) {
                Either.Right(dto.itemsList.stream().map { TradeSummaryItemConverter.toModel(it) }.collect(Collectors.toList()))
            } else {
                Either.Right(Collections.emptyList())
            }
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }

    }

    fun getSummaryMatrix(supportMetadata: SupportMetadata, sharesOnly: Boolean): Either<ApiException, List<SummaryMatrixModel>> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/transactions/summary/matrix/")
                .queryParam("shares-only", sharesOnly)

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        return try {
            val responseEntity = restTemplate
                    .exchange(builder.build("").toString(), HttpMethod.GET, HttpEntity<Any>(headers), TLSummaryMatrixResponse::class.java)

            val dto = responseEntity.body

            return if (dto != null) {
                Either.Right(dto.itemsList.stream().map { SummaryMatrixConverter.toModel(it) }.collect(Collectors.toList()))
            } else {
                Either.Right(Collections.emptyList())
            }
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }
    }

    fun createShareTransaction(supportMetadata: SupportMetadata, model: ShareJournalModel): Either<ApiException, ShareJournalModel> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/shares")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        val requestDto = ShareJournalConverter.toTLDto(model)

        return try {
            val responseEntity = restTemplate
                    .exchange(builder.build("").toString(), HttpMethod.POST, HttpEntity<Any>(requestDto, headers), TLShareJournalDto::class.java)

            val responseDto: TLShareJournalDto? = responseEntity.body

            return if (responseDto != null) {
                Either.Right(ShareJournalConverter.toModel(responseDto))
            } else {
                Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_DATA_CONVERSION_ERROR, "Cannot process the response"))
            }
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }

    }

    fun editShareTransaction(supportMetadata: SupportMetadata, model: ShareJournalModel): Either<ApiException, Unit> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/shares/{id}")

        val headers = HttpHeaders()

        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        val requestDto = ShareJournalConverter.toTLDto(model)

        return try {
            restTemplate
                    .exchange(builder.build(model.transactionId).toString(), HttpMethod.PUT, HttpEntity<Any>(requestDto, headers), Any::class.java)
            Either.Right(Unit)
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }
    }

    fun deleteShareTransaction(supportMetadata: SupportMetadata, transactionId: String): Either<ApiException, Unit> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/shares/{id}")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        return try {
            restTemplate.exchange(builder.build(transactionId).toString(), HttpMethod.DELETE, HttpEntity<Any>(headers), Any::class.java)
            Either.Right(Unit)
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }
    }

    fun createOptionTransaction(supportMetadata: SupportMetadata, model: OptionJournalModel): Either<ApiException, OptionJournalModel> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/options")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        val requestDto = OptionJournalConverter.toTLDto(model)

        return try {
            val responseEntity = restTemplate
                    .exchange(builder.build("").toString(), HttpMethod.POST, HttpEntity<Any>(requestDto, headers), TLOptionJournalDto::class.java)

            val responseDto: TLOptionJournalDto? = responseEntity.body

            return if (responseDto != null) {
                Either.Right(OptionJournalConverter.toModel(responseDto))
            } else {
                Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_DATA_CONVERSION_ERROR, "Cannot process the response"))
            }
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }

    }

    fun editOptionTransaction(supportMetadata: SupportMetadata, model: OptionJournalModel): Either<ApiException, Unit> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/options/{id}")

        val headers = HttpHeaders()

        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        val requestDto = OptionJournalConverter.toTLDto(model)
        return try {
            restTemplate
                    .exchange(builder.build(model.transactionId).toString(), HttpMethod.PUT, HttpEntity<Any>(requestDto, headers), Any::class.java)
            Either.Right(Unit)
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }
    }

    fun deleteOptionTransaction(supportMetadata: SupportMetadata, transactionId: String): Either<ApiException, Unit> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/options/{id}")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        return try {
            restTemplate.exchange(builder.build(transactionId).toString(), HttpMethod.DELETE, HttpEntity<Any>(headers), Any::class.java)
            Either.Right(Unit)
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }
    }

    fun updateTransactionSettings(supportMetadata: SupportMetadata, model: TransactionSettingsModel): Either<ApiException, Unit> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("transactions/settings/{id}")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        val requestDto = TransactionSettingsConverter.toDto(model)

        return try {
            restTemplate.exchange(builder.build(model.transactionId).toString(), HttpMethod.PUT, HttpEntity<Any>(requestDto, headers), Any::class.java)
            Either.Right(Unit)
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }
    }

    fun createDividendTransaction(supportMetadata: SupportMetadata, model: DividendJournalModel): Either<ApiException, DividendJournalModel> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/dividends")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        val requestDto = DividendJournalConverter.toTLDto(model)

        return try {
            val responseEntity = restTemplate
                    .exchange(builder.build("").toString(), HttpMethod.POST, HttpEntity<Any>(requestDto, headers), TLDividendJournalDto::class.java)

            val responseDto: TLDividendJournalDto? = responseEntity.body

            if (responseDto != null) {
                Either.Right(DividendJournalConverter.toModel(responseDto))
            } else {
                Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_DATA_CONVERSION_ERROR, "Cannot process the response"))
            }
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }

    }

    fun deleteDividendTransaction(supportMetadata: SupportMetadata, transactionId: String): Either<ApiException, Unit> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/dividends/{id}")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, supportMetadata.accountId.toString())
        headers.set(PORTFOLIO_ID_PARAM_NAME, supportMetadata.portfolioId.toString())

        return try {
            restTemplate.exchange(builder.build(transactionId).toString(), HttpMethod.DELETE, HttpEntity<Any>(headers), Any::class.java)
            Either.Right(Unit)
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }
    }

    fun getDefaultPortfolio(accountId: UUID): Either<ApiException, PortfolioModel> {
        val builder = UriComponentsBuilder
                .fromHttpUrl(url)
                .path("/portfolios/default/")

        val headers = HttpHeaders()
        headers.set("Content-Type", PROTOBUF_MEDIA_TYPE_VALUE)
        headers.set(ACCOUNT_ID_HEADER_NAME, accountId.toString())

        return try {
            val responseEntity = restTemplate.exchange(builder.build("").toString(), HttpMethod.GET, HttpEntity<Any>(headers), TLPortfolioDto::class.java)
            val responseDto: TLPortfolioDto? = responseEntity.body

            if (responseDto != null) {
                Either.Right(PortfolioConverter.toModel(responseDto))
            } else {
                Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_DATA_CONVERSION_ERROR, "Cannot process the response"))
            }
        } catch (rce: RestClientException) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_CONNECTION_ERROR, "Could not connect to TradingLogApi"))
        } catch (ex: Exception) {
            Either.Left(ApiException(ApiExceptionCode.EXTERNAL_API_OTHER_ERROR, ex.message))
        }
    }
}
